src/garage/tf/algos/batch_polopt2.py

```diff
@@ -0,0 +1,281 @@
+    def __init__(self,
+                 env_spec,
+                 policy,
+                 baseline,
+                 scope=None,
+                 max_path_length=500,
+                 discount=0.99,
+                 gae_lambda=1,
+                 center_adv=True,
+                 positive_adv=False,
+                 fixed_horizon=False,
+                 flatten_input=True):
+        self.env_spec = env_spec
+        self.policy = policy
+        self.baseline = baseline
+        self.scope = scope
+        self.max_path_length = max_path_length
+        self.discount = discount
+        self.gae_lambda = gae_lambda
+        self.center_adv = center_adv
+        self.positive_adv = positive_adv
+        self.fixed_horizon = fixed_horizon
+        self.flatten_input = flatten_input
+
+        self.episode_reward_mean = collections.deque(maxlen=100)
+        if policy.vectorized:
+            self.sampler_cls = OnPolicyVectorizedSampler
+        else:
+            self.sampler_cls = BatchSampler
+
+        self.init_opt()
+
+    def train(self, runner):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for _ in runner.step_epochs():
+            runner.step_path = runner.obtain_samples(runner.step_itr)
+            last_return = self.train_once(runner.step_itr, runner.step_path)
+            runner.step_itr += 1
+
+        return last_return
+
+    def train_once(self, itr, paths):
+        """Perform one step of policy optimization given one batch of samples.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            float: The average return in last epoch cycle.
+
+        """
+        paths = self.process_samples(itr, paths)
+        self.log_diagnostics(paths)
+        logger.log('Optimizing policy...')
+        self.optimize_policy(itr, paths)
+        return paths['average_return']
+
+    def log_diagnostics(self, paths):
+        """Log diagnostic information.
+
+        Args:
+            paths (list[dict]): A list of collected paths.
+
+        """
+        logger.log('Logging diagnostics...')
+        self.policy.log_diagnostics(paths)
+        self.baseline.log_diagnostics(paths)
+
+    def process_samples(self, itr, paths):
+        """Return processed sample data based on the collected paths.
+
+        Args:
+            itr (int): Iteration number.
+            paths (list[dict]): A list of collected paths.
+
+        Returns:
+            dict: Processed sample data, with key
+                * observations  : (numpy.ndarray), shape [B * (T), *obs_dims]
```

Comment:
I appreciate the extra effort on formatting, but unfortunately I think the documentation generator doesn't used fixed-width, so these won't align when they're rendered.

I think you should stick to the style used for args.

If you want to present something as fixed width (e.g. the shapes), you can use `backticks` just like in markdown

Reply:
How about if I put it in the docstring body instead of under returns?

Reply:
sure. that doesn't change what i said about spacing, though.